onnxruntime/core/providers/webgpu/math/gemm_utils.cc

```diff
@@ -272,22 +272,35 @@ Status MakeMatMulPackedVec4Source(ShaderHelper& shader,
       << "const rowPerThread = " << elements_per_thread_y << ";\n"
       << "const colPerThread = " << elements_per_thread_x << ";\n"
       << "const innerElementSize = " << inner_elements_size << ";\n"
-      << "const tileInner = " << tile_inner << ";\n";
+      << "const tileInner = " << tile_inner << ";\n"
+      << "const workgroup_size = vec3u(workgroup_size_x, workgroup_size_y, workgroup_size_z);\n";
+
+  // Compute `logical_workgroup_id` and `logical_global_id` because the dispatch workgroup size in
+  // `ProgramBase.SetDispatchGroupSize()` may be normalized in
+  // `ProgramManager::NormalizeDispatchGroupSize()`. In the shader we should always use
+  // `logical_workgroup_id` and `logical_global_id` instead of `workgroup_id` and `global_id`.
+  shader.MainFunctionBody()
+      << "  let logical_workgroups_xy = uniforms.logical_dispatch_x * uniforms.logical_dispatch_y;\n"
+      << "  let logical_workgroup_id_z = workgroup_idx / logical_workgroups_xy;\n"
+      << "  let logical_workgroup_id_y = (workgroup_idx % logical_workgroups_xy) / uniforms.logical_dispatch_x;\n"
+      << "  let logical_workgroup_id_x = (workgroup_idx % logical_workgroups_xy) % uniforms.logical_dispatch_x;\n"
+      << "  let logical_workgroup_id = vec3u(logical_workgroup_id_x, logical_workgroup_id_y, logical_workgroup_id_z);\n"
+      << "  let logical_global_id = logical_workgroup_id * workgroup_size + local_id;\n";
```

Comment:
nit: For code reusing, maybe it's better to wrap line 283- line 288 into a string variable and reuse it in line 520-525.